[BUG] Card Browser: "Can't sort notes by Custom"

![image](https://github.com/ankidroid/Anki-Android/assets/62114487/5620a476-2c3d-4405-a225-482a9318580b)

```
W  com.ichi2.libanki.exception.InvalidSearchException: net.ankiweb.rsdroid.exceptions.BackendInvalidInputException: Can't sort Notes by Custom.
  	at com.ichi2.libanki.Collection.findNotes(Collection.kt:398)
  	at com.ichi2.libanki.Collection.findOneCardByNote(Collection.kt:411)
  	at com.ichi2.anki.CardBrowserKt.searchForCards$lambda$0(CardBrowser.kt:2311)
  	at com.ichi2.anki.CardBrowserKt.$r8$lambda$SvxaASTuq0hKmRmVoFHOLFGl_0Y(Unknown Source:0)
  	at com.ichi2.anki.CardBrowserKt$$ExternalSyntheticLambda2.invoke(D8$$SyntheticClass:0)
  	at com.ichi2.anki.CollectionManager.withCol$lambda$1(CollectionManager.kt:116)
  	at com.ichi2.anki.CollectionManager.$r8$lambda$ffr4nUlMOyAvThaKA2WuJlnLaPU(Unknown Source:0)
  	at com.ichi2.anki.CollectionManager$$ExternalSyntheticLambda10.invoke(D8$$SyntheticClass:0)
  	at com.ichi2.anki.CollectionManager$withQueue$3.invokeSuspend(CollectionManager.kt:100)
  	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
  	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
  	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
  	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:100)
  	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
  	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
  	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
  	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
  Caused by: net.ankiweb.rsdroid.exceptions.BackendInvalidInputException: Can't sort Notes by Custom.
  	at net.ankiweb.rsdroid.exceptions.BackendInvalidInputException$Companion.fromInvalidInputError(BackendInvalidInputException.kt:37)
  	at net.ankiweb.rsdroid.BackendException$Companion.fromError(BackendException.kt:114)
  	at net.ankiweb.rsdroid.BackendKt.unpackResult(Backend.kt:271)
  	at net.ankiweb.rsdroid.BackendKt.access$unpackResult(Backend.kt:1)
  	at net.ankiweb.rsdroid.Backend$runMethodRaw$1.invoke(Backend.kt:118)
  	at net.ankiweb.rsdroid.Backend$runMethodRaw$1.invoke(Backend.kt:117)
  	at net.ankiweb.rsdroid.Backend.withBackend(Backend.kt:131)
  	at net.ankiweb.rsdroid.Backend.runMethodRaw(Backend.kt:117)
  	at anki.backend.GeneratedBackend.searchNotesRaw(GeneratedBackend.kt:1847)
  	at anki.backend.GeneratedBackend.searchNotes(GeneratedBackend.kt:1853)
  	at com.ichi2.libanki.Collection.findNotes(Collection.kt:396)
  	at com.ichi2.libanki.Collection.findOneCardByNote(Collection.kt:411) 
  	at com.ichi2.anki.CardBrowserKt.searchForCards$lambda$0(CardBrowser.kt:2311) 
  	at com.ichi2.anki.CardBrowserKt.$r8$lambda$SvxaASTuq0hKmRmVoFHOLFGl_0Y(Unknown Source:0) 
  	at com.ichi2.anki.CardBrowserKt$$ExternalSyntheticLambda2.invoke(D8$$SyntheticClass:0) 
  	at com.ichi2.anki.CollectionManager.withCol$lambda$1(CollectionManager.kt:116) 
  	at com.ichi2.anki.CollectionManager.$r8$lambda$ffr4nUlMOyAvThaKA2WuJlnLaPU(Unknown Source:0) 
  	at com.ichi2.anki.CollectionManager$$ExternalSyntheticLambda10.invoke(D8$$SyntheticClass:0) 
  	at com.ichi2.anki.CollectionManager$withQueue$3.invokeSuspend(CollectionManager.kt:100) 
  	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
  	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106) 
  	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115) 
  	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:100) 
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584) 
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793) 
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697) 
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684) 
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt
@@ -303,7 +303,7 @@ class CardBrowserViewModel(
             flowOfCardsOrNotes.update { cardsOrNotes }
 
             withCol {
-                sortTypeFlow.update { SortType.fromCol(config, sharedPrefs()) }
+                sortTypeFlow.update { SortType.fromCol(config, cardsOrNotes, sharedPrefs()) }
                 reverseDirectionFlow.update { ReverseDirection.fromConfig(config) }
             }
             Timber.i("initCompleted")

--- a/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt
@@ -88,7 +88,8 @@ class CardBrowserViewModel(
     private val lastDeckIdRepository: LastDeckIdRepository,
     private val cacheDir: File,
     options: CardBrowserLaunchOptions?,
-    preferences: SharedPreferencesProvider
+    preferences: SharedPreferencesProvider,
+    private val manualInit: Boolean = false
 ) : ViewModel(), SharedPreferencesProvider by preferences {
 
     // Set by the UI to determine the number of cards to preload before returning search results
@@ -306,8 +307,17 @@ class CardBrowserViewModel(
                 reverseDirectionFlow.update { ReverseDirection.fromConfig(config) }
             }
             Timber.i("initCompleted")
-            flowOfInitCompleted.update { true }
+
+            if (!manualInit) {
+                flowOfInitCompleted.update { true }
+            }
         }
+    }
+
+    @VisibleForTesting
+    fun manualInit() {
+        require(manualInit) { "'manualInit' should be true" }
+        flowOfInitCompleted.update { true }
     }
 
     /** Whether any rows are selected */

--- a/AnkiDroid/src/main/java/com/ichi2/anki/model/SortType.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/model/SortType.kt
@@ -55,6 +55,8 @@ enum class SortType(val ankiSortType: String?, val cardBrowserLabelIndex: Int) {
         config.set("sortType", this.ankiSortType ?: SORT_FIELD.ankiSortType)
         config.set("noteSortType", this.ankiSortType ?: SORT_FIELD.ankiSortType)
         preferences.edit {
+            // TODO: This should be changed to use the collection
+            // and have a different value for cards & notes
             putBoolean("cardBrowserNoSorting", this@SortType == NO_SORTING)
         }
     }
@@ -64,8 +66,9 @@ enum class SortType(val ankiSortType: String?, val cardBrowserLabelIndex: Int) {
         if (this == NO_SORTING) SortOrder.NoOrdering() else SortOrder.UseCollectionOrdering()
 
     companion object {
-        fun fromCol(config: Config, preferences: SharedPreferences): SortType {
-            val colOrder = config.get<String>("sortType")
+        fun fromCol(config: Config, cardsOrNotes: CardsOrNotes, preferences: SharedPreferences): SortType {
+            val configKey = if (cardsOrNotes == CardsOrNotes.CARDS) "sortType" else "noteSortType"
+            val colOrder = config.get<String>(configKey)
             val type = entries.firstOrNull { it.ankiSortType == colOrder } ?: NO_SORTING
             if (type == SORT_FIELD && preferences.getBoolean("cardBrowserNoSorting", false)) {
                 return NO_SORTING

--- a/AnkiDroid/src/test/java/com/ichi2/anki/browser/CardBrowserViewModelTest.kt
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/browser/CardBrowserViewModelTest.kt
@@ -206,7 +206,7 @@ class CardBrowserViewModelTest : JvmTest() {
         }
     }
 
-    private fun runViewModelTest(notes: Int = 0, testBody: suspend CardBrowserViewModel.() -> Unit) = runTest {
+    private fun runViewModelTest(notes: Int = 0, manualInit: Boolean = true, testBody: suspend CardBrowserViewModel.() -> Unit) = runTest {
         for (i in 0 until notes) {
             addNoteUsingBasicModel()
         }
@@ -214,8 +214,13 @@ class CardBrowserViewModelTest : JvmTest() {
             lastDeckIdRepository = SharedPreferencesLastDeckIdRepository(),
             cacheDir = createTransientDirectory(),
             options = null,
-            preferences = AnkiDroidApp.sharedPreferencesProvider
+            preferences = AnkiDroidApp.sharedPreferencesProvider,
+            manualInit = manualInit
         )
+        // makes ignoreValuesFromViewModelLaunch work under test
+        if (manualInit) {
+            viewModel.manualInit()
+        }
         testBody(viewModel)
     }
 

--- a/AnkiDroid/src/test/java/com/ichi2/anki/browser/CardBrowserViewModelTest.kt
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/browser/CardBrowserViewModelTest.kt
@@ -206,6 +206,17 @@ class CardBrowserViewModelTest : JvmTest() {
         }
     }
 
+    @Test
+    fun `sort order from notes is selected - 16514`() {
+        col.config.set("sortType", "noteCrt")
+        col.config.set("noteSortType", "_field_Frequency")
+        with(col) { CardsOrNotes.NOTES.saveToCollection() }
+
+        runViewModelTest(notes = 1) {
+            assertThat("1 row returned", rowCount, equalTo(1))
+        }
+    }
+
     private fun runViewModelTest(notes: Int = 0, manualInit: Boolean = true, testBody: suspend CardBrowserViewModel.() -> Unit) = runTest {
         for (i in 0 until notes) {
             addNoteUsingBasicModel()

